### PR TITLE
Decorators for deprecation and named arguments validation

### PR DESCRIPTION
**Pull request by qubvel**
# What does this PR do?

Introduces 2 decorators for a better approach to deprecate and validate named arguments.

## Deprecation decorator

Function or method decorator to notify users about deprecated keyword arguments, replacing them with a new name if specified.

This decorator allows you to:
 - Notify users when a keyword argument is deprecated.
 - Automatically replace deprecated keyword arguments with new ones.
 - Raise an error if deprecated arguments are used, depending on the specified conditions.

By default, the decorator notifies the user about the deprecated argument while the `transformers.__version__`  < specified `version` in the decorator. To keep notifications with any version `warn_if_greater_or_equal_version=True` can be set.

#### Example usage **with** renaming argument:

```python
from transformers.utils.deprecation import deprecate_kwarg

@deprecate_kwarg("reduce_labels", new_name="do_reduce_labels", version="6.0.0")
def my_function(do_reduce_labels):
    print(do_reduce_labels)

my_function(reduce_labels=True)  # Will show a deprecation warning and use do_reduce_labels=True
```
```output
FutureWarning: `reduce_labels` is deprecated and will be removed in version 6.0.0 for `my_function`. Use `do_reduce_labels` instead.
```

#### Example when both deprecated and new arguments were specified for function call:

```python
from transformers.utils.deprecation import deprecate_kwarg

@deprecate_kwarg("reduce_labels", new_name="do_reduce_labels", version="6.0.0")
def my_function(do_reduce_labels):
    print(do_reduce_labels)

my_function(do_reduce_labels=False, reduce_labels=True)  # not deprecated kwarg has higher priority
# False
```
```output
FutureWarning: Both `reduce_labels` and `do_reduce_labels` are set for `my_function`. Using `do_reduce_labels=False` and ignoring deprecated `reduce_labels=True`.
  my_function(do_reduce_labels=False, reduce_labels=True)  # not deprecated kwarg has higher priority
```

#### Example usage **without** renaming argument:

```python
from transformers.utils.deprecation import deprecate_kwarg

@deprecate_kwarg("max_size", version="6.0.0")
def my_function(max_size):
    print(max_size)

my_function(max_size=1333)  # Will show a deprecation warning
```
```output
FutureWarning: `max_size` is deprecated and will be removed in version 6.0.0 for `my_function`.
  my_function(max_size=1333)  # Will show a deprecation warning
  ```
  
## Validation decorator

Decorator to filter out named arguments that are not in the function signature.
This decorator ensures that only the keyword arguments that match the function's signature, or are specified in the
`extra` list, are passed to the function. Any additional keyword arguments are filtered out and a **warning is issued**.

Example usage:

```python
from transformers.utils.generic import filter_out_non_signature_kwargs

@filter_out_non_signature_kwargs(extra=["allowed_extra_arg"])
def my_function(arg1, arg2, **kwargs):
    print("Function args:", arg1, arg2, kwargs)

my_function(arg1=1, arg2=2, allowed_extra_arg=3, invalid_arg=4)
```
```output
UserWarning: The following named arguments are not valid for `my_function` and were ignored: 'invalid_arg'
  my_function(arg1=1, arg2=2, allowed_extra_arg=3, invalid_arg=4)
Function args: 1 2 {'allowed_extra_arg': 3}
```

## Additional testing

Introducing such decorators might be tricky and lead to some unexpected issues since historically all **kwargs arguments were captured and saved as image_processor's attributes. Despite testing the decorators themselves and tests that passed for image processors, additional testing was conducted for configurations on the hub.

For each image processor in ["maskformer", "mask2former", "segformer", "beit", "oneformer"], configs from the hub with more than 10 downloads were pulled, and the result of the preprocess method was compared with the main branch to ensure they are identical.

Code just for reference:
https://gist.github.com/qubvel/85cf17f516063d63219ef773a365e83d

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/beit/image_processing_beit.py [hunk 1]
```diff
@@ -109,12 +108,7 @@ def __init__(
         **kwargs,
     ) -> None:
         if "reduce_labels" in kwargs:
-            warnings.warn(
-                "The `reduce_labels` parameter is deprecated and will be removed in a future version. Please use"
-                " `do_reduce_labels` instead.",
-                FutureWarning,
-            )
-            do_reduce_labels = kwargs.pop("reduce_labels")
+            raise ValueError("The `reduce_labels` parameter has been deprecated. Use `do_reduce_labels` instead.")
```
Comment (amyeroberts) [inline src/transformers/models/beit/image_processing_beit.py:111]: Sadly we can't do this directly for two reasons: 

* A deprecation version wasn't specified in the message (my bad)
* This is going to break a lot of things for users

We absolutely want to push users to try and use the correct argument, and we should update this value on all of the official checkpoints. However, because this is from the config, which users might not have control over or know how to change; and there might be hundreds of configs with the old value on and off the hub, we're at risk of breaking code and losing users.

What I'd suggest here is: 
* We update the warning message to specify a specific version (typically two versions after the release version this commit will be included in). 
* After this version, we remove the warning. We silently handle this to maintain backwards compatibility i.e. popping the value, and make sure all official references to the flag are removed; we don't guarantee any future maintenance for this value.

File: src/transformers/models/mask2former/image_processing_mask2former.py [hunk 2]
```diff
@@ -424,6 +424,13 @@ def __init__(
             self._max_size = kwargs.pop("max_size")
         else:
             self._max_size = 1333
+        if "reduce_labels" in kwargs:
+            warnings.warn(
+                "The `reduce_labels` parameter is deprecated and will be removed in a future version. Please use"
+                " `do_reduce_labels` instead.",
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/image_processing_mask2former.py:430]: We should specify a version here 
```suggestion
                "The `reduce_labels` parameter is deprecated and will be removed in v4.43. Please use"
                " `do_reduce_labels` instead.",
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for working on this - great to see more alignment between the image processors. 

Only comment is to not raise an exception if reduce_labels is passed in

File: src/transformers/models/beit/image_processing_beit.py (hunk 1 above)
Comment (qubvel) [reply src/transformers/models/beit/image_processing_beit.py:111]: Yes, that's breaking changes...

The idea was to support reading old configs overriding `from_dict` method with
```python
@classmethod
    def from_dict(cls, image_processor_dict: Dict[str, Any], **kwargs):
    ...
    if "reduce_labels" in image_processor_dict:
            image_processor_dict["do_reduce_labels"] = image_processor_dict.pop("reduce_labels")
```
Here we silently replace `reduce_labels` with `do_reduce_labels`.

By raising an exception in `__init__`, we prevent users from creating **new** `image_processor` with deprecated parameters. So that might break some fine-tuning scripts, which the user should have control of.

Does it make sense, or am I missing some use cases?
Are such kinds of breaking changes possible only over major versions like v4->v5?

Comment (amyeroberts) [reply src/transformers/models/beit/image_processing_beit.py:111]: > Here we silently replace reduce_labels with do_reduce_labels.

This is great - we definitely want to do this! It means any old config, if saved out, will be fully compatible. 

> Are such kinds of breaking changes possible only over major versions like v4->v5?

Kind of, it's more to do with the type and scale of breaking change, and the controls users have over it. 

We do deprecate methods, flags, etc. within the codebase semi-regularly. And we do tolerate a certain amount of breaking changes. For example, in musicgen recently, training was enabled. This technically created a breaking change, as one of the models previously returned a loss. The update meant that anyone using the model would see the loss suddenly change. However, as the previous loss was incorrect and the model use relatively low this was deemed OK. 

There's two main reasons why we wouldn't want to raise an exception here at the moment:
* Ability to change the code behaviour. Users don't always have the option of updating the config: they might not have permissions to modify on the hub.
* Awareness of configs. Many users are not familiar with the configs. It's easier to deprecate calls to e.g. `object.method(foo, bar)` as we emit a warning, and the user just removes that line of code. Changing the config is less obvious. 

The reason I think this is because when we first tried to deprecate flags for the image processors, we got quite a few messages from users either complaining or confused about why they were happening. I'm yet to see a single config updated because of them!

It is something we could do in v5, although we may not want to.

Comment (qubvel) [reply src/transformers/models/beit/image_processing_beit.py:111]: Thanks a lot for the clarification, I will summarize PR just to make sure we are on the same page.

There are 3 most popular ways to create an image processor with deprecated parameters (sorted by popularity, high->low, IMHO)

| Method                                                                                                                        | Is BC in current PR                    |
|-------------------------------------------------------------------------------------------------------------------------------|-----------------------------------------------|
| image_processor = ImageProcessor.from_pretrained(<br>&emsp;    "repo/model_with_deprecated_params"<br>)                             | ✅ Yes<br>Because the user does not<br> have control of it (by overriding `from_dict()`)          |
| image_processor = ImageProcessor.from_pretrained(<br>&emsp;    "repo/model_with_deprecated_params",<br>&emsp;    reduce_labels=True,<br>) | 🟥 No, raise error<br>Because it is explicit parameter setting |
| image_processor = ImageProcessor(<br>&emsp;    ...<br>&emsp;    reduce_labels=True,<br>)                                                  | 🟥 No, raise error<br>Because it is explicit parameter setting |
|                                                                                                                               |                                               |

I hope this approach:
1. Pushes the user to update their codebase, which they have control over, while still allowing using previously created models.
2. Will allow us to gradually move backward compatibility for deprecated params only in one method `from_dict` (ideally), making code easier to support.
3. Probably, we can also get rid of `**kwargs`, have a strict set of arguments, and replace checks with a decorator for renamed/removed arguments:
```python
@depecated("reduce_labels", replace_with="do_reduce_labels", from="4.42.0", to="4.45.0")
def preprocess(...):
```
The decorator will rename the argument and raise a warning in between, and then we can remove it with automated checks, or leave it for a while to raise an explicit error.
At the moment, because of `**kwargs`, anyone might easily make a typo in an argument name and it will be silently passed, leading to unexpected behavior that is hard to debug.
As a simple example:
```python
from transformers import DetrImageProcessor

image_processor = DetrImageProcessor.from_pretrained(
    "facebook/detr-resnet-50",
    do_nomalize=False, <----- here set as False
)
print(image_processor)

# DetrImageProcessor {
# ...
#   "do_convert_annotations": true,
#   "do_normalize": true, <----- but we get True
#   "do_pad": true, 
#   "do_rescale": true,
#   "do_resize": true,
# ...
# }
```

We can test this strategy with some particular models, not super popular, just to see how it works.

I might overestimate the problem or underestimate risks, user behavior, or edge cases. You have much more experience maintaining the library, so let me know if any of these thoughts are valuable. And thank you once again for your clarifications with examples, it is much easier to follow 🤗

Comment (amyeroberts) [reply src/transformers/models/beit/image_processing_beit.py:111]: 100% aligned with you comment above, thanks for taking the time to write this up so clearly. 

I really like the idea of a decorator. Let's ask @molbap his opinion on this, as he's been handling a lot of these considerations across processors and image processor.

Comment (molbap) [reply src/transformers/models/beit/image_processing_beit.py:111]: OOO for a few days but checked and it's an interesting take. I like the decorator idea too. I wholeheartedly agree with
>  At the moment, because of **kwargs, anyone might easily make a typo in an argument name and it will be silently passed, leading to unexpected behavior that is hard to debug.

That is almost always true. That's why getting rid of **kwargs is impossible, as we need to be able to handle arbitrary inputs without breaking runtime in situations where people have non-canonical configs.

Hence the point of having a list of valid arguments: that way, we can at least raise a warning and list the arguments passed by a user that are not doing anything. In your case, if `do_normalize` is part of the valid image processor arguments but `do_nomalize` (typo) is not, then we can raise a clear warning saying that `do_nomalize` isn't used anywhere. 
We can even push this a bit and do a quick levenshtein check conditioning a `perhaps you made a typo?` message. That might be too much.

For the decorator option to mark args as deprecated, LGTM too! Less loc, and forces us to write a version number. I'd say the "`from`" is not necessarily useful as we want to encourage using the latest version, imo. 
Thanks for looking into this @qubvel! Kind of related is https://github.com/huggingface/transformers/pull/30511 that I want to merge ASAP, didn't have time to get back to it with but it's closely connected to this problem.

Comment (qubvel) [reply src/transformers/models/beit/image_processing_beit.py:111]: Thanks for looking into this @molbap 

Probably we can use a function signature in decorator to filter not-used kwargs instead of passing them.
What do you think about the following?

```python
import functools
import inspect
import warnings
import numpy as np


def filter_not_used_arguments(func):
    @functools.wraps(func)
    def wrapper(*args, **kwargs):

        sig = inspect.signature(func)
        valid_params = sig.parameters.keys()
        valid_kwargs = {k: v for k, v in kwargs.items() if k in valid_params}
        invalid_params = {k: v for k, v in kwargs.items() if k not in valid_params}
        
        if invalid_params:
            warnings.warn(f"Unused named arguments: {', '.join(invalid_params.keys())}")
        
        return func(*args, **valid_kwargs)
    
    return wrapper


class ImageProcessor:

    @filter_not_used_arguments
    def preprocess(self, image, do_normalize=False, do_reduce_labels=False):
        pass


image_processor = ImageProcessor()
image = np.ones((100, 100, 3))

# passing invalid  `do_nomalize` instead of `do_normalize`
image_processor.preprocess(image, do_nomalize=True, do_reduce_labels=True)  
# UserWarning: Unused named arguments: do_nomalize
```

1. This will simplify validation, we will not need to store the list of available arguments
2. Make validation method-specific instead of class-specific
3. While this is backward compatible and we do not raise an error here, IDE will help users and force users to update parameters if no **kwargs is specified for the function
Screenshot 2024-05-20 at 10 36 32

Comment (molbap) [reply src/transformers/models/beit/image_processing_beit.py:111]: That looks like a great solution :) looks very good to me

**Comment by qubvel** [comment]
@molbap @amyeroberts I made a new iteration here. Added two decorators
1) `@deprecate_kwarg("reduce_labels", new_name="do_reduce_labels", version="4.41.0")`
 - rename argument if `new_name` is provided
 - raise warning if current version < `version` specified,  then just do it silently
 2) `filter_out_non_signature_kwargs(extra=["max_size"])`
  - filter out all named args that are not in the function signature and not specified in `extra`. As mentioned by @molbap it is not always possible to get rid of all `**kwargs` (for example, it might include complex logic of parameters interaction).
  
 Please, have a look, feedback is appreciated 🤗

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation.py [hunk 3]
```diff
@@ -108,7 +108,7 @@ class DataTrainingArguments:
             )
         },
     )
-    reduce_labels: Optional[bool] = field(
+    do_reduce_labels: Optional[bool] = field(
```
Comment (amyeroberts) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation.py:111]: For the script, we'll need to deprecate `reduce_labels` e.g. [like here](https://github.com/huggingface/transformers/blob/aada568f73953d0678bbb0d605bfeed46fb209fd/examples/flax/token-classification/run_flax_ner.py#L399)

File: examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py [hunk 4]
```diff
@@ -86,7 +86,7 @@ def parse_args():
         default="segments/sidewalk-semantic",
     )
     parser.add_argument(
-        "--reduce_labels",
+        "--do_reduce_labels",
```
Comment (amyeroberts) [inline examples/pytorch/semantic-segmentation/run_semantic_segmentation_no_trainer.py:89]: Same here re deprecation

File: src/transformers/utils/deprecation.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    warn(message, FutureWarning)
+
+
+def deprecate_kwarg(
+    old_name: str,
+    version: str,
+    new_name: Optional[str] = None,
+    raise_if_ge_version: bool = False,
+    raise_if_both_names: bool = False,
+    add_message: Optional[str] = None,
```
Comment (amyeroberts) [inline src/transformers/utils/deprecation.py:26]: nit - could we call this `additional_message`? `add_message` reads as a bool to me (don't feel strongly about this, so happy if you decide to stick with the original)

File: src/transformers/utils/deprecation.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+                message = f"`{old_name}` is deprecated {version_message}."
+
+            if message is not None and add_message is not None:
+                message = f"{message} {add_message}"
+
+            # raise error or notify user
+            if minimum_action == Action.RAISE:
+                raise ValueError(message)
+            elif minimum_action == Action.NOTIFY and raise_if_ge_version and is_already_deprecated:
+                raise ValueError(message)
```
Comment (amyeroberts) [inline src/transformers/utils/deprecation.py:83]: nit - I'd separate this out between setting the minimum action and the raise/notify logic, just to make it a bit clearer

```suggestion
            if minimum_action == Action.NOTIFY and raise_if_ge_version and is_already_deprecated:
                miminum_action = Action.RAISE
            
            # raise error or notify user
            if minimum_action == Action.RAISE:
                raise ValueError(message)
```

File: src/transformers/utils/kwargs_validation.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
+import functools
+import inspect
+import warnings
+from typing import Optional
+
+
+def filter_out_non_signature_kwargs(extra: Optional[list] = None):
```
Comment (amyeroberts) [inline src/transformers/utils/kwargs_validation.py:7]: ❤️

**Review by amyeroberts — COMMENTED** [review]
Really beautiful work - this could be very useful across the library and will help kill so much repeated code ❤️  

Only major comment is about the new files src/transformers/utils/kwargs_validation.py and src/transformers/utils/deprecation.py, and whether we want to add these whole new modules into utils. 

Personally, I think `src/transformers/utils/deprecation.py` makes sense. For `filter_out_non_signature_kwargs`, I'd place it under (the admittedly badly named) `utils/generic.py` for now

Would be great to get a second opinion on this from @molbap!

**Review by molbap — COMMENTED** [review]
I like the kwargs validation method, so much cleaner! Left a few comments :)

File: src/transformers/utils/kwargs_validation.py [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    """Filter out named arguments that are not in the function signature."""
+
+    extra = extra or []
+    extra_params_to_pass = set(extra)
+
+    def decorator(func):
+        sig = inspect.signature(func)
+        function_named_args = set(sig.parameters.keys())
+        valid_kwargs_to_pass = function_named_args.union(extra_params_to_pass)
+        is_class_method = "self" in function_named_args
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/utils/kwargs_validation.py:17]: `is_class_method` is probably not the best name here :) This attribute is only used to get class name for better warning raising. Decorator itself can be used for static, class methods, and functions.
I will also add a check for "cls" attribute.

Here is how the warning looks for different methods, class name is shown for instance and class methods, but not for static.

```python
from transformers.utils.kwargs_validation import filter_out_non_signature_kwargs

class MyClass:

    @filter_out_non_signature_kwargs()
    def __init__(self, a, b, c):
        self.a = a
        self.b = b
        self.c = c

    @filter_out_non_signature_kwargs()
    @staticmethod
    def static_method_1(a, b, c):
        return a, b, c

    @staticmethod
    @filter_out_non_signature_kwargs()
    def static_method_2(a, b, c):
        return a, b, c

    @classmethod
    @filter_out_non_signature_kwargs()
    def class_method_1(cls, a, b, c):
        return a, b, c
    
    ## This one will throw an error during class construction
    ## The only way to use it is to define @classmethod on top of @filter_out_non_signature_kwargs()
    # @filter_out_non_signature_kwargs()
    # @classmethod
    # def class_method_2(cls, a, b, c):
    #     return a, b, c

@filter_out_non_signature_kwargs()
def regular_function(a, b, c):
    return a, b, c


a = MyClass(1, 2, 3, d=4)
# ...: UserWarning: The following named arguments are not valid for `MyClass.__init__` and were ignored: d

MyClass.static_method_1(1, 2, 3, e=4)  # @filter_out_non_signature_kwargs, then @staticmethod
# ...: UserWarning: The following named arguments are not valid for `static_method_1` and were ignored: e

MyClass.static_method_2(1, 2, 3, f=4) # @staticmethod, then @filter_out_non_signature_kwargs
# ...: UserWarning: The following named arguments are not valid for `static_method_2` and were ignored: f

MyClass.class_method_1(1, 2, 3, g=4)
# ...: UserWarning: The following named arguments are not valid for `MyClass.class_method_1` and were ignored: g

regular_function(1, 2, 3, h=4)
# ...: UserWarning: The following named arguments are not valid for `regular_function` and were ignored: h
```

File: src/transformers/utils/deprecation.py [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+
+def warning(message):
+    warn(message, FutureWarning)
+
+
+def deprecate_kwarg(
+    old_name: str,
+    version: str,
+    new_name: Optional[str] = None,
+    raise_if_ge_version: bool = False,
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/utils/deprecation.py]: The motivation was to keep it short to keep the decorator in one line but probably will be rarely used since we prefer not to raise errors for deprecated args.

File: src/transformers/models/segformer/image_processing_segformer.py [hunk 10]
```diff
@@ -119,33 +112,15 @@ def __init__(
...
-            "do_rescale",
-            "rescale_factor",
-            "do_normalize",
-            "image_mean",
-            "image_std",
-            "do_reduce_labels",
-            "return_tensors",
-            "data_format",
-            "input_data_format",
-        ]
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/segformer/image_processing_segformer.py:137]: Not sure I got it, could you please provide an example or more details?

[the remark this answers is not in the document]
Comment (molbap) [reply src/transformers/models/segformer/image_processing_segformer.py:137]: Hey, sure! What I mean is that the validation decorator is built by 
```python
        sig = inspect.signature(func)
        function_named_args = set(sig.parameters.keys())
        valid_kwargs_to_pass = function_named_args.union(extra_params_to_pass)
```

Hence for SegFormer, where 

```python
    @deprecate_kwarg("reduce_labels", new_name="do_reduce_labels", version="4.41.0")
    @filter_out_non_signature_kwargs()
    def __init__(
        self,
        do_resize: bool = True,
        size: Dict[str, int] = None,
        resample: PILImageResampling = PILImageResampling.BILINEAR,
        do_rescale: bool = True,
        rescale_factor: Union[int, float] = 1 / 255,
        do_normalize: bool = True,
        image_mean: Optional[Union[float, List[float]]] = None,
        image_std: Optional[Union[float, List[float]]] = None,
        do_reduce_labels: bool = False,
    ) -> None:
...
```

Here we're using the init signature, and it seems to match, but in `vit_hybrid` (for instance) you could miss kwargs like `input_data_format` that are _not_ part of the _init_ signature, but are present in `preprocess`

```python
    def __init__(
        self,
        do_resize: bool = True,
        size: Dict[str, int] = None,
        resample: PILImageResampling = PILImageResampling.BICUBIC,
        do_center_crop: bool = True,
        crop_size: Dict[str, int] = None,
        do_rescale: bool = True,
        rescale_factor: Union[int, float] = 1 / 255,
        do_normalize: bool = True,
        image_mean: Optional[Union[float, List[float]]] = None,
        image_std: Optional[Union[float, List[float]]] = None,
        do_convert_rgb: bool = True,
        **kwargs,
    ) -> None:
        super().__init__(**kwargs)
        size = size if size is not None else {"shortest_edge": 224}
        size = get_size_dict(size, default_to_square=False)
        crop_size = crop_size if crop_size is not None else {"height": 224, "width": 224}
        crop_size = get_size_dict(crop_size, default_to_square=True, param_name="crop_size")

        self.do_resize = do_resize
        self.size = size
        self.resample = resample
        self.do_center_crop = do_center_crop
        self.crop_size = crop_size
        self.do_rescale = do_rescale
        self.rescale_factor = rescale_factor
        self.do_normalize = do_normalize
        self.image_mean = image_mean if image_mean is not None else OPENAI_CLIP_MEAN
        self.image_std = image_std if image_std is not None else OPENAI_CLIP_STD
        self.do_convert_rgb = do_convert_rgb
        self._valid_processor_keys = [
            "images",
            "do_resize",
            "size",
            "resample",
            "do_center_crop",
            "crop_size",
            "do_rescale",
            "rescale_factor",
            "do_normalize",
            "image_mean",
            "image_std",
            "do_convert_rgb",
            "return_tensors",
            "data_format",
            "input_data_format",
        ]
```

Hope that clarifies the previous message!

[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/models/segformer/image_processing_segformer.py:137]: Thanks for the clarification, @molbap!

The decorator should be applied per method, specifically for both `__init__` and `preprocess`.

Regarding vit_hybrid, I noticed that `input_data_format` is present in the preprocess method signature, so it will be handled correctly by the decorator.

However, there could be a potential issue **if** the preprocess method uses `self.input_data_format` that was not defined in the `__init__ `signature and was instead captured by passing `**kwargs` to `super()`. I'm not certain if we have instances where we reference an attribute not defined in `__init__`, but these cases should be managed accurately.

Probably, it is worth writing a short script that will pull the top 100 most popular `image_processor` configs for each model and compare "main" vs "current branch" methods where decorators were applied.

File: src/transformers/models/beit/image_processing_beit.py [hunk 11]
```diff
@@ -14,12 +14,11 @@
 # limitations under the License.
 """Image processor class for Beit."""
 
-import warnings
 from typing import Any, Dict, List, Optional, Tuple, Union
 
 import numpy as np
 
-from ...image_processing_utils import BaseImageProcessor, BatchFeature, get_size_dict
+from ...image_processing_utils import INIT_SERVICE_KWARGS, BaseImageProcessor, BatchFeature, get_size_dict
```
Comment (qubvel) [inline src/transformers/models/beit/image_processing_beit.py:21]: Restored from config image processors propogate `image_processor_type` and `processor_type` named arguments to `__init__` method. Despite they are not necessary and image_processor can be initialized without these arguments, some test logic relies on them. I tried to remove them, but it looks like too many models should be touched, probably, it's better to try as a follow-up. 
For now, I came up with this suboptimal solution, just to specify them as extra names for `@filter_out_non_signature_kwargs(extra=INIT_SERVICE_KWARGS)`

**Comment by qubvel** [comment]
I modified PR's name and description to make it clear.
@amyeroberts @molbap could you please have a look once again, it should be done now.

File: src/transformers/utils/deprecation.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        @wraps(func)
+        def wrapped_func(*args, **kwargs):
+            minimum_action = None
+            message = None
+
+            # deprecated kwarg and its new version are set for function call -> replace it with new name
+            if old_name in kwargs and new_name in kwargs:
+                minimum_action = Action.RAISE if raise_if_both_names else Action.NOTIFY
+                message = f"Both `{old_name}` and `{new_name}` are set. Using `{new_name}={kwargs[new_name]}` and ignoring deprecated `{old_name}={kwargs[old_name]}`."
+                kwargs.pop(old_name)
```
Comment (molbap) [inline src/transformers/utils/deprecation.py:115]: I was wondering about this specific use-case, I may be missing something: if I do

```python
proc = BeitImageProcessor(reduce_labels=True, weird_arg=True)
# UserWarning: The following named arguments are not valid for `BeitImageProcessor.__init__` and were ignored: 'weird_arg'
```

I'm getting a warning here, but if I call the two versions of the argument (which can be strings it seems, but that's unrelated)

```python
proc = BeitImageProcessor(reduce_labels="any value really", do_reduce_labels="another value")
print( proc(images=np_img)['pixel_values'][0].shape )
# (3, 224, 224)
# no warning here
```
but it seems it should be the same log level?

File: src/transformers/models/oneformer/image_processing_oneformer.py [hunk 13]
```diff
@@ -473,26 +470,30 @@ def __init__(
...
 
+    @classmethod
+    def from_dict(cls, image_processor_dict: Dict[str, Any], **kwargs):
+        """
+        Overrides the `from_dict` method from the base class to save support of deprecated `reduce_labels` in old configs
+        """
+        image_processor_dict = image_processor_dict.copy()
+        if "reduce_labels" in image_processor_dict:
+            image_processor_dict["do_reduce_labels"] = image_processor_dict.pop("reduce_labels")
+        return super().from_dict(image_processor_dict, **kwargs)
```
Comment (molbap) [inline src/transformers/models/oneformer/image_processing_oneformer.py:483]: this is great, is there a way to abstract this to a generic method too?

**Review by molbap — COMMENTED** [review]
Again, very clean - not sure about the image_processor_type and processor_class, maybe @amyeroberts will have a suggestion! Added a few comments but other than that lgtm :+1:

File: src/transformers/utils/deprecation.py (hunk 12 above)
Comment (qubvel) [reply src/transformers/utils/deprecation.py:115]: Oh, this is a nice catch! Will fix and add a test!

Comment (qubvel) [reply src/transformers/utils/deprecation.py]: addressed in https://github.com/huggingface/transformers/pull/30799/commits/9156618dc3b0aaf1e68460aa6f7e43f83e781700

File: src/transformers/models/oneformer/image_processing_oneformer.py (hunk 13 above)
Comment (qubvel) [reply src/transformers/models/oneformer/image_processing_oneformer.py]: I would prefer to remove class specific `from_dict` method since now the decorator handling this, but for that, we have to update `super().from_dict`, which is currently separate `image_processor_dict` and `**kwargs` assignment for image processor initialization.

 - `image_processor_dict` is passed to `__init__` of image processor
 - `**kwargs` are assigned to image_processor instance with `image_processor.key = value`
 
 I addressed it in https://github.com/huggingface/transformers/pull/30799/commits/2b21c88f65e41b6b971afb77e4376b0c98d26d19, however, there would be a small breaking change:

Previously:
 - ✅ `**kwargs` have higher priority over `image_processor_dict`.
 - 🔴 deprecated kwarg in `**kwargs` has higher priority then **not** deprecated one in `image_processor_dict`

Updated: 
 - ✅ `**kwargs` have higher priority over `image_processor_dict`.
 - 🔴 **not** deprecated argument has higher priority no matter where it was specified either in `image_processor_dict` or in `**kwargs`
 
 For example:
 
 ```python
 image_processor_dict = {"do_reduce_labels": True}

 # Before
image_processor = MaskFormerImageProcessor.from_dict(image_processor_dict, reduce_labels=False)
image_processor.do_reduce_labels == False  # reduce_labels deprecated but still has higher priority
 
 # Now
image_processor = MaskFormerImageProcessor.from_dict(image_processor_dict, reduce_labels=False)

# reduce_labels deprecated -> do_reduce_labels from `image_processor_dict` has higher priority 
# + WARNING will be issued
image_processor.do_reduce_labels == True  


 ```

File: tests/utils/test_generic.py [hunk 14]
```diff
@@ -198,3 +200,74 @@ def test_expand_dims_flax(self):
         x = np.random.randn(3, 4)
         t = jnp.array(x)
         self.assertTrue(np.allclose(expand_dims(x, axis=1), np.asarray(expand_dims(t, axis=1))))
+
+
+class ValidationDecoratorTester(unittest.TestCase):
```
Comment (amyeroberts) [inline tests/utils/test_generic.py:205]: Nice :)

File: src/transformers/models/beit/image_processing_beit.py (hunk 11 above)
Comment (amyeroberts) [reply src/transformers/models/beit/image_processing_beit.py:21]: I think this a pretty clean and pragmatic solution

**Review by amyeroberts — APPROVED** [review]
Beautiful piece of work 🔥 It's great to see so much repeated code cleaned up - this should make managing the deprecation for all of the processing classes a lot easier!

File: src/transformers/utils/generic.py [hunk 15]
```diff
@@ -750,3 +751,79 @@ def infer_framework(model_class):
...
+
+        my_function(arg1=1, arg2=2, allowed_extra_arg=3, invalid_arg=4)
+        # This will print: 1 2 {"allowed_extra_arg": 3}
+        # And issue a warning: "The following named arguments are not valid for `my_function` and were ignored: 'invalid_arg'"
+        ```
+    """
+    extra = extra or []
+    extra_params_to_pass = set(extra)
+
+    def decorator(func):
```
[the remark this answers is not in the document]
Comment (qubvel) [reply src/transformers/utils/generic.py:786]: Should we add type hints in such cases? It`s just an internal function inside the decorator and not going to be used somewhere else.

File: src/transformers/models/oneformer/image_processing_oneformer.py (hunk 13 above)
Comment (molbap) [reply src/transformers/models/oneformer/image_processing_oneformer.py]: I see! thanks for the clear explanation. Then, I think it's ok to have a class-specific method :)

**Comment by qubvel** [comment]
Rebase + remove generic from_dict (discussed here https://github.com/huggingface/transformers/pull/30799#discussion_r1631283521)
